### PR TITLE
feat(config): add `diff` option

### LINKED ISSUE #2401: Custom snapshot console colors (A11Y)
### Clear and concise description of the problem

* Colors work differently in different editors/consoles/terminals/shells/CI's/etc.
* Some people have different visual, or cognitive impairments that make certain color combinations frustrating or difficult to discern.

### Suggested solution

The default red/green approach is fine by me, and probably good for the majority of users. But it would be nice if there was a `vite.config.js` option for customizing the colors used by Vitest, particularly in snapshot failures.

```js
import chalk from 'chalk';
import { defineConfig } from 'vite';

export default defineConfig({
  test: {
    diff: {
      // I'm not sure what the API would look like here, So I'll just use what
      // jest-diff documents for this example. The following is copied directly from
      // https://github.com/facebook/jest/tree/main/packages/jest-diff#options
      aAnnotation: 'Expected',
      aColor: chalk.green,
      aIndicator: '-',
      bAnnotation: 'Received',
      bColor: chalk.red,
      bIndicator: '+',
      changeColor: chalk.inverse,
      changeLineTrailingSpaceColor: string => string,
      commonColor: chalk.dim,
      commonIndicator: ' ',
      commonLineTrailingSpaceColor: string => string,
      compareKeys: undefined,
      contextLines: 5,
      emptyFirstOrLastLinePlaceholder: '',
      expand: true,
      includeChangeCounts: false,
      omitAnnotationLines: false,
      patchColor: chalk.yellow
    }
  }
});
```

By exposing all options, it empowers communities to create their own commonly recommended settings, like so:

```js
import { defineConfig } from 'vite';
import diff from 'vitest-deuteranopia-diffs';

export default defineConfig({ test: { diff } });
```
```js
import { defineConfig } from 'vite';
import diff from 'vitest-cognitive-diffs';

export default defineConfig({ test: { diff } });
```
```js
import { defineConfig } from 'vite';
import diff from 'vitest-win-cmd-diffs';

export default defineConfig({ test: { diff } });
```
etc.
There could even be an `awesome-vitest-diffs` repo with screenshots of each library, so users can more easily choose what works best for them or their teams. This could all be community-lead, but we need Vitest to allow for this.

### Alternative

_No response_

### Additional context

This same feature request was submitted to Jest, and approved for PR.

* https://github.com/facebook/jest/issues/12576

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by fenghan34**
### Description

Closes #2401 

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [x] Run the tests with `pnpm test:ci`.

### Documentation
- [x] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [x] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

File: packages/runner/src/collect.ts [hunk 1]
```diff
@@ -33,6 +33,7 @@ export async function collectTests(paths: string[], runner: VitestRunner): Promi
     try {
       const setupStart = now()
       await runSetupFiles(config, runner)
+      await loadDiffOptionFile(config, runner)
```
Comment (fenghan34) [inline packages/runner/src/collect.ts:36]: Not sure if here is the right place to load diff config. Could we do this earlier? 🤔

File: packages/runner/src/types/tasks.ts [hunk 2]
```diff
@@ -259,6 +260,11 @@ export interface TestContext {
    */
   task: Readonly<Test>
 
+  /**
+   * Diff options
+   */
+  diff?: DiffOptions
```
Comment (sheremet-va) [inline packages/runner/src/types/tasks.ts:266]: Why is it on a task?

File: packages/runner/src/setup.ts [hunk 3]
```diff
@@ -15,3 +15,13 @@ export async function runSetupFiles(config: VitestRunnerConfig, runner: VitestRu
...
 }
+
+export async function loadDiffOptionFile(config: VitestRunnerConfig, runner: VitestRunner) {
+  if (typeof config.diff !== 'string')
+    return
+
+  const diffModule = await runner.importFile(config.diff, 'diff') as any
+
+  if (diffModule && diffModule.default)
+    runner.config.diff = diffModule.default
```
Comment (sheremet-va) [inline packages/runner/src/setup.ts:26]: Should have more checks and throw an error if it's not on default for example

File: packages/runner/src/context.ts [hunk 4]
```diff
@@ -49,6 +50,7 @@ export function createTestContext(test: Test, runner: VitestRunner): TestContext
 
   context.meta = test
   context.task = test
+  context.diff = runner.config.diff as DiffOptions
```
Comment (sheremet-va) [inline packages/runner/src/context.ts:53]: Why do we pollute context here?

File: packages/runner/src/collect.ts (hunk 1 above)
Comment (sheremet-va) [reply packages/runner/src/collect.ts:36]: Is there a reason why you load it for each test file separately when there is only one config?

**Review by sheremet-va — CHANGES_REQUESTED** [review]

File: packages/runner/src/collect.ts (hunk 1 above)
Comment (fenghan34) [reply packages/runner/src/collect.ts:36]: Looks like I totally misunderstood "It should be loaded inside a test context" 😅

File: packages/runner/src/types/tasks.ts (hunk 2 above)
Comment (fenghan34) [reply packages/runner/src/types/tasks.ts:266]: My mistake, just realized we can access it by `runner.config`

File: packages/runner/src/collect.ts (hunk 1 above)
Comment (fenghan34) [reply packages/runner/src/collect.ts:36]: Could you tell me where should I load this config file? Thanks!

Comment (sheremet-va) [reply packages/runner/src/collect.ts:36]: > inside a test context

It just means that we need to load it in the test runtime, not on the server side. I can recommend doing it in setup.common and then just pass this config down to the runner

Comment (fenghan34) [reply packages/runner/src/collect.ts]: Now I got it, thanks for your patience and guidance!

File: packages/vitest/src/runtime/entry-vm.ts [hunk 5]
```diff
@@ -41,7 +41,9 @@ export async function run(files: string[], config: ResolvedConfig, executor: Vit
   if (config.chaiConfig)
     setupChaiConfig(config.chaiConfig)
 
+  const diffConfig = await loadDiffConfig(config, executor)
```
Comment (sheremet-va) [inline packages/vitest/src/runtime/entry-vm.ts:44]: Can't you do this inside `resolveTestRunner`?

**Review by sheremet-va — CHANGES_REQUESTED** [review]
I don't see this option working in the browser?

File: packages/vitest/src/runtime/entry-vm.ts (hunk 5 above)
Comment (fenghan34) [reply packages/vitest/src/runtime/entry-vm.ts]: Thank you for reviewing, already updated.

**Comment by fenghan34** [comment]
> I don't see this option working in the browser?

I wasn't aware of that, just added browser mode support.

File: docs/config/index.md [hunk 6]
```diff
@@ -1619,3 +1619,11 @@ By default Vitest will run all of your test cases even if some of them fail. Thi
...
 
 Retry the test specific number of times if it fails.
+
+### diff
+
+- **Type:** `string`
+- **CLI:** `--diff=<value>`
+- **Version:** Since Vitest 0.34.4
+
+Path to a diff config that will be used to generate diff interface. Useful if you want to customize diff display.
```
Comment (sheremet-va) [inline docs/config/index.md:1629]: Can you add a code example here? To showcase how to define it using typescript. A little bit of information how colors are defined can also be useful (need start/end property I think).

**Pull request merged**